### PR TITLE
bgp: BGP Control Plane modularization

### PR DESCRIPTION
**Pull request by dylandreimerink**
This work converts the BGP Control Plane controller and BGP Route Manager into hive cells, leaving as much of the existing code intact. These cells are now hooked into the agent hive directly. The daemon now takes the Controller as parameter both to preserve the behavior of setting the controller as a field value on the daemon and so the BGP controllers lifecycle events are invoked.

This is the first PR in a series of prep work for SVC announcements. Future PRs will break up the single cell into more granular pieces which should improve the test ability and extensibility of the BGP Control Plane.

### CODE REVIEW

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [inline pkg/bgpv1/agent/controller.go:189]: Curious about this, wouldn't it be possible to simply not add the BGP-Controller cell into the Controller hive, if we detect its not enabled in the config?

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [inline pkg/bgpv1/agent/controller.go]: Is there any reason not to do this in the `Start` hook? https://github.com/cilium/cilium/pull/22183/files#diff-663c8aba9492de183ebe045c8bf425d6a6a645970cd2a201a7073e9b21ffe8adR272

This just feels a bit better as you're creating the context where the runtime will set the controller into action.

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [inline pkg/bgpv1/agent/controller.go]: My gut wants Run() to take a ctx here.

File: pkg/bgpv1/agent/nodespecer.go
Comment (ldelossa) [inline pkg/bgpv1/agent/nodespecer.go:32]: Why exactly these changes? Is this just code refactoring?

**Review by ldelossa — CHANGES_REQUESTED** [review]
First run thru, some questions.

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [reply pkg/bgpv1/agent/controller.go:189]: I played with this idea early on, but decided against it as it would’ve made inspecting the object graph harder and the implementation more complex (providers would need to be constructors of constructors).

I feel like it’s good enough to return nil or no-op objects when things are disabled.

File: pkg/bgpv1/agent/controller.go
Comment (dylandreimerink) [reply pkg/bgpv1/agent/controller.go]: Adopted this pattern due to feedback from other reviewers on #21764. I can see an argument for doing it both ways, will move it to `Start()`.

File: pkg/bgpv1/agent/controller.go
Comment (dylandreimerink) [reply pkg/bgpv1/agent/controller.go]: Yea, this again comes down to `context.Context` usage patterns. I know you prefer passing via parameter, so I will refactor to that pattern.

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [reply pkg/bgpv1/agent/controller.go]: Consider whether [workerpool](https://github.com/cilium/workerpool) would be a good fit here. E.g. instead of `go c.Run()` etc. you'd do `c.workerpool.Submit("run", c.Run)` and on Stop() you'd `c.workerpool.Close()` which cancels the context passed to each worker and waits for them to finish.

File: pkg/bgpv1/agent/nodespecer.go
Comment (dylandreimerink) [reply pkg/bgpv1/agent/nodespecer.go:32]: Yes. We are moving the starting of the informer from the BGP CP constructor to the nodespecer run function. Since nodespecer is consuming the informer, it seems logical that it should own this reference.

File: pkg/bgpv1/agent/controller.go
Comment (dylandreimerink) [reply pkg/bgpv1/agent/controller.go]: Yes, I think it is, given we have 3 seperate go routines in Start() as well

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go:189]: gotchya.

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go]: Agreed, becomes cleaner.

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go]: Gracias.

File: pkg/bgpv1/agent/nodespecer.go
Comment (ldelossa) [reply pkg/bgpv1/agent/nodespecer.go:32]: Okay i think that makes sense.

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [inline pkg/bgpv1/agent/controller.go]:
```suggestion
	RouterMgr     BGPRouterManager
```

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [inline pkg/bgpv1/agent/controller.go]: While this was in the original code, I don't think this is a useful thing to log. Hive already logs all start hook calls, so that should be enough.

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [inline pkg/bgpv1/agent/controller.go:283]: This should return `ctx.Err()` (maybe augmented with "workerpool didn't close in time") as otherwise it hides the fact that `c.workerpool.Close()` took too long and leaves resources running in the background. E.g. tests might continue onto other tests or benchmarks that might be affected by those resources.

In my code I've not bothered to handle it this way. I've just called `c.workerpool.Close()` directly and ignore the hook context as I can't use it to cancel the stop.

Could perhaps consider just dropping the stop context altogether since it's rarely useful. Start context I think is useful as it allows extracting maximum amount information when the start times out rather than just fataling. And again for max information we want the agent to itself abort the start instead of having e.g. k8s kill it due to failing health checks.

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [inline pkg/bgpv1/agent/controller.go:177]: Hmm, just thinking out loud:

Can we already introduce a BGPConfig (without Flags()) and do a projection from DaemonConfig into it?
Or since we have the `ControllerOpt` stuff already, could we create a set of those from `DaemonConfig`?
E.g. try to avoid exposing the whole DaemonConfig into the innards of modules.

File: pkg/bgpv1/agent/controller.go
Comment (joamaki) [inline pkg/bgpv1/agent/controller.go]: Ah, ControllerOpt isn't used at all. I don't think there's a reason to "future proof" this as only users of this are within the same repository, so I would suggest dropping it altogether.

**Review by joamaki — APPROVED** [review]

File: pkg/bgpv1/agent/controller.go
Comment (dylandreimerink) [reply pkg/bgpv1/agent/controller.go]: Yes, the `ControllerOpt`s were already in place, just refactored it to the new pattern. It might be for enterprise, @ldelossa what is your take on this?

File: pkg/bgpv1/agent/controller.go
Comment (dylandreimerink) [reply pkg/bgpv1/agent/controller.go:177]: As discussed offline, we can completely remove the `EnableBGPControlPlane` from the DaemonConfig and handle it via the cell.Config. Would like to do that in the followup PR since this PR still uses the `IPAM` field as well, which is no longer required in the next PR.

File: pkg/bgpv1/agent/controller.go
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go]: Not used. The API is just stamped out for future options. 

It seems plausible that as features are added to the controller, there will be options you'll want to apply to it. 

But I guess, it can just be added back at that time. 

It's not used in enterprise neither.

File: pkg/bgpv1/cell.go
Comment (thorn3r) [inline pkg/bgpv1/cell.go]: tiny nit: typo here
```suggestion
	// implemented, provide the manager via a Cell that picks implementation based on configuration.
```

**Review by thorn3r — APPROVED** [review]
LGTM :+1:

**Review by ldelossa — APPROVED** [review]
LGTM

**Comment by dylandreimerink** [comment]
Alright got 2 approvals. Ignoring doc-structure for now since they were added due to a doc commit added to the wrong branch on accident (since been resolved).

Almost all tests are green, except for some flakes:
* ConformanceAKS is failing in a test to connect to cloudflare `to-fqdns/pod-to-world/https-to-one-one-one-one-index-1: cilium-test/client-769787f6f8-n9flc (10.0.0.66) -> one-one-one-one-https-index (one.one.one.one:443)` which is a trend at the moment.
* TravisCI is hitting #22224 
* test-1.16-4.9 and test-1.25-net-next are both hitting #22404. This is a new flake related to TLS policy which seems unrelated to this work.
*  runtime (test-runtime) hits the same flake as travis #22224

**Review by jrajahalme — APPROVED** [review]

**Comment by jrajahalme** [comment]
@dylandreimerink @joestringer Unfortunaltely this broke Travis CI like below, not sure why I did not see this before merging:
```
=== RUN   Test
level=info msg="Creating etcd client" ConfigPath= KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=error msg="Invoke failed" ="cmd.glob..func2 (daemon_main.go:1627)" error="could not build arguments for function \"github.com/cilium/cilium/daemon/cmd\".glob..func2 (/home/ubuntu/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1627): failed to build promise.Promise[*github.com/cilium/cilium/daemon/cmd.Daemon]: could not build arguments for function \"github.com/cilium/cilium/daemon/cmd\".newDaemonPromise (/home/ubuntu/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1642): failed to build *agent.Controller: missing dependencies for function \"github.com/cilium/cilium/pkg/bgpv1/agent\".NewController (/home/ubuntu/go/src/github.com/cilium/cilium/pkg/bgpv1/agent/controller.go:188): missing type: *option.DaemonConfig" subsys=hive

----------------------------------------------------------------------
FAIL: daemon_test.go:223: DaemonEtcdSuite.SetUpTest

daemon_test.go:224:
    e.DaemonSuite.SetUpTest(c)
daemon_test.go:157:
    c.Assert(err, IsNil)
... value dig.errArgumentsFailed = could not build arguments for function "github.com/cilium/cilium/daemon/cmd".glob..func2 (/home/ubuntu/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1627): failed to build promise.Promise[*github.com/cilium/cilium/daemon/cmd.Daemon]: could not build arguments for function "github.com/cilium/cilium/daemon/cmd".newDaemonPromise (/home/ubuntu/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1642): failed to build *agent.Controller: missing dependencies for function "github.com/cilium/cilium/pkg/bgpv1/agent".NewController (/home/ubuntu/go/src/github.com/cilium/cilium/pkg/bgpv1/agent/controller.go:188): missing type: *option.DaemonConfig
```
As you see this has to do with the new hive thing that I don't know much about, but apparently some of the changes made on the production code would also need to be reflected in the test suite.

To reproduce this locally, you can:
```
make start-kvstores
cd daemon/cmd
go test -v -tags=integration_tests ./...
```

I verified this by reverting this commit and trying again, and the test passes.

**Pull request merged**
